File: 200-number-of-islands/number-of-islands.cs
public class Solution {
    public int NumIslands(char[][] grid) {
        if (grid == null || grid.Length == 0) {
            return 0;
        }

        int numIslands = 0;
        int m = grid.Length;
        int n = grid[0].Length;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    numIslands++;
                    DFS(grid, i, j, m, n);
                }
            }
        }

        return numIslands;
    }

    private void DFS(char[][] grid, int i, int j, int m, int n) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == '0') {
            return;
        }

        grid[i][j] = '0'; // Mark this cell as visited by setting it to '0'

        // Explore all four possible directions (up, down, left, right)
        DFS(grid, i + 1, j, m, n); // Down
        DFS(grid, i - 1, j, m, n); // Up
        DFS(grid, i, j + 1, m, n); // Right
        DFS(grid, i, j - 1, m, n); // Left
    }
}

File: 200-number-of-islands/number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    if (!grid || grid.length === 0) return 0;

    let count = 0;
    let rows = grid.length;
    let cols = grid[0].length;

    const dfs = (i, j) => {
        if (i < 0 || j < 0 || i >= rows || j >= cols || grid[i][j] === '0') {
            return;
        }

        grid[i][j] = '0'; 
        dfs(i - 1, j);
        dfs(i + 1, j); 
        dfs(i, j - 1);
        dfs(i, j + 1); 
    };

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (grid[i][j] === '1') {
                count++;
                dfs(i, j);
            }
        }
    }

    return count;
};
